BLD: threads.h with MacOS 14.x ARM + gcc 14.2.0

Another thing observed while investigating gh-21826--if building with `gcc 14.2.0` toolchain (from `homebrew`) on MacOS ARM  (OS version `14.7`, alongside OpenBLAS), I see an issue with `fatal error: threads.h: No such file or directory`.

Incantation on latest SciPy `main`: `CC=gcc-14 CXX=g++-14 python dev.py build -j 16`

Build error:

```
<snip>
[129/1274] Generating scipy/linalg/flapack_module with a custom command
[130/1274] Compiling C object scipy/linalg/_flapack.cpython-312-darwin.so.p/meson-generated_..__flapackmodule.c.o
FAILED: scipy/linalg/_flapack.cpython-312-darwin.so.p/meson-generated_..__flapackmodule.c.o 
gcc-14 -Iscipy/linalg/_flapack.cpython-312-darwin.so.p -Iscipy/linalg -I../scipy/linalg -I../../../python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/numpy/_core/include -I../../../python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/numpy/f2py/src -I/opt/homebrew/Cellar/openblas/0.3.28/include -I/Library/Frameworks/Python.framework/Versions/3.12/include/python3.12 -fvisibility=hidden -fdiagnostics-color=always -Wall -Winvalid-pch -std=c17 -O2 -g -Wno-unused-but-set-variable -Wno-unused-function -Wno-conversion -Wno-misleading-indentation -Wno-empty-body -MD -MQ scipy/linalg/_flapack.cpython-312-darwin.so.p/meson-generated_..__flapackmodule.c.o -MF scipy/linalg/_flapack.cpython-312-darwin.so.p/meson-generated_..__flapackmodule.c.o.d -o scipy/linalg/_flapack.cpython-312-darwin.so.p/meson-generated_..__flapackmodule.c.o -c scipy/linalg/_flapackmodule.c
scipy/linalg/_flapackmodule.c:90:10: fatal error: threads.h: No such file or directory
   90 | #include <threads.h>
      |          ^~~~~~~~~~~
compilation terminated.

<snip>
```

There is a complicated NumPy issue about this kind of thing here for example: https://github.com/numpy/numpy/issues/24761

--- a/scipy/_lib/meson.build
+++ b/scipy/_lib/meson.build
@@ -31,10 +31,12 @@ lib_cython_gen = generator(cython,
   output : '@BASENAME@.c',
   depends : [_cython_tree, _lib_pxd])
 
+ccallback_dep = declare_dependency(include_directories: ['..', 'src'])
+
 py3.extension_module('_ccallback_c',
   lib_cython_gen.process('_ccallback_c.pyx'),
   c_args: [cython_c_args, Wno_discarded_qualifiers],
-  include_directories: 'src',
+  dependencies: ccallback_dep,
   link_args: version_link_args,
   install: true,
   subdir: 'scipy/_lib'
@@ -42,7 +44,7 @@ py3.extension_module('_ccallback_c',
 
 py3.extension_module('_test_ccallback',
   'src/_test_ccallback.c',
-  include_directories: 'src',
+  dependencies: ccallback_dep,
   link_args: version_link_args,
   install: true,
   subdir: 'scipy/_lib',

--- a/scipy/_lib/src/ccallback.h
+++ b/scipy/_lib/src/ccallback.h
@@ -22,6 +22,8 @@
 #include <Python.h>
 #include <setjmp.h>
 
+#include "scipy_config.h"
+
 /* Default behavior */
 #define CCALLBACK_DEFAULTS 0x0
 /* Whether calling ccallback_obtain is enabled */
@@ -66,32 +68,9 @@ struct ccallback {
  * Thread-local storage
  */
 
-#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && (__GNUC_MINOR__ >= 4)))
+#if !defined(SCIPY_TLS_EMPTY)
 
-static __thread ccallback_t *_active_ccallback = NULL;
-
-static void *ccallback__get_thread_local(void)
-{
-    return (void *)_active_ccallback;
-}
-
-static int ccallback__set_thread_local(void *value)
-{
-    _active_ccallback = value;
-    return 0;
-}
-
-/*
- * Obtain a pointer to the current ccallback_t structure.
- */
-static ccallback_t *ccallback_obtain(void)
-{
-    return (ccallback_t *)ccallback__get_thread_local();
-}
-
-#elif defined(_MSC_VER)
-
-static __declspec(thread) ccallback_t *_active_ccallback = NULL;
+static SCIPY_TLS ccallback_t *_active_ccallback = NULL;
 
 static void *ccallback__get_thread_local(void)
 {

--- a/scipy/integrate/meson.build
+++ b/scipy/integrate/meson.build
@@ -63,8 +63,7 @@ dop_lib = static_library('dop_lib',
 
 py3.extension_module('_quadpack',
   ['__quadpack.h', '__quadpack.c'],
-  include_directories: ['../_lib/src'],
-  dependencies: np_dep,
+  dependencies: [np_dep, ccallback_dep],
   install: true,
   subdir: 'scipy/integrate'
 )

--- a/scipy/meson.build
+++ b/scipy/meson.build
@@ -136,11 +136,51 @@ fortranobject_dep = declare_dependency(
   link_with: fortranobject_lib,
   include_directories: [inc_np, inc_f2py],
 )
-# Workaround for numpy#24761 on numpy<1.26.1 (see also gh-20515)
-_f2py_c_args = []
-if is_mingw
-  _f2py_c_args = '-DNPY_OS_MINGW'
+
+cdata = configuration_data()
+
+# Test variable attribute to use for thread-local storage;
+# Adapted from `numpy/_core/meson.build`.
+check_tls_attrs = [
+  ['thread_local', 'HAVE_THREAD_LOCAL'],    # C23
+  ['_Thread_local', 'HAVE__THREAD_LOCAL'],  # C11/C17
+  ['__thread', 'HAVE__THREAD'],
+]
+if is_windows and not is_mingw
+  check_tls_attrs += ['__declspec(thread)', 'HAVE___DECLSPEC_THREAD_']
 endif
+f2py_tls_define = ''
+foreach tls_attrs: check_tls_attrs
+  attr = tls_attrs[0]
+  code = f'''
+    #pragma GCC diagnostic error "-Wattributes"
+    #pragma clang diagnostic error "-Wattributes"
+
+    int @attr@ foo;
+  '''
+  code += '''
+    int
+    main()
+    {
+      return 0;
+    }
+  '''
+  if cc.compiles(code, name: tls_attrs[0])
+    cdata.set10(tls_attrs[1], true)
+    f2py_tls_define = tls_attrs[0]
+  endif
+endforeach
+
+# Contains only TLS check results for now - name chosen for when more compiler
+# checks need adding in the future.
+scipy_config_h = configure_file(
+  input: 'scipy_config.h.in',
+  output: 'scipy_config.h',
+  configuration: cdata,
+  install: false
+)
+
+_f2py_c_args = [f'-DF2PY_THREAD_LOCAL_DECL=@f2py_tls_define@']
 fortranobject_dep = declare_dependency(
   dependencies: fortranobject_dep,
   compile_args: _f2py_c_args,

--- a/scipy/ndimage/meson.build
+++ b/scipy/ndimage/meson.build
@@ -9,8 +9,8 @@ py3.extension_module('_nd_image',
     'src/ni_splines.c',
     'src/ni_support.c'
   ],
-  include_directories: ['../_lib/src', '../_build_utils/src'],
-  dependencies: np_dep,
+  include_directories: ['../_build_utils/src'],
+  dependencies: [np_dep, ccallback_dep],
   link_args: version_link_args,
   install: true,
   subdir: 'scipy/ndimage'

--- a/scipy/optimize/meson.build
+++ b/scipy/optimize/meson.build
@@ -19,8 +19,7 @@ py3.extension_module('_minpack',
     '__minpack.h',
     '__minpack.c',
   ],
-  include_directories: '../_lib/src',
-  dependencies: np_dep,
+  dependencies: [np_dep, ccallback_dep],
   install: true,
   subdir: 'scipy/optimize'
 )

--- a/scipy/scipy_config.h.in
+++ b/scipy/scipy_config.h.in
@@ -1,0 +1,19 @@
+/* Thread-local storage */
+#mesondefine HAVE_THREAD_LOCAL
+#mesondefine HAVE__THREAD_LOCAL
+#mesondefine HAVE__THREAD
+#mesondefine HAVE___DECLSPEC_THREAD_
+
+#ifdef __cplusplus
+    #define SCIPY_TLS thread_local
+#elif defined(HAVE_THREAD_LOCAL)
+    #define SCIPY_TLS thread_local
+#elif defined(HAVE__THREAD_LOCAL)
+    #define SCIPY_TLS _Thread_local
+#elif defined(HAVE___THREAD)
+    #define SCIPY_TLS __thread
+#elif defined(HAVE___DECLSPEC_THREAD_)
+    #define SCIPY_TLS __declspec(thread)
+#else
+    #define SCIPY_TLS
+#endif

--- a/scipy/stats/_unuran/meson.build
+++ b/scipy/stats/_unuran/meson.build
@@ -156,7 +156,6 @@ unuran_sources = [
 ]
 
 unuran_include_dirs = [
-  '../../_lib/src',  # for messagestream.h
   '../../_lib/unuran/unuran/src',
   '../../_lib/unuran/unuran/src/distr',
   '../../_lib/unuran/unuran/src/distributions',
@@ -228,7 +227,7 @@ py3.extension_module('unuran_wrapper',
   ],
   c_args: [unuran_defines, cython_c_args],
   include_directories: [unuran_include_dirs],
-  dependencies: np_dep,
+  dependencies: [np_dep, ccallback_dep],
   link_args: version_link_args,
   install: true,
   subdir: 'scipy/stats/_unuran'

